File: Probelm2.java
//Time complexity:O(N X 4^N)
//Space copmplexity:O(log base 3 (length of string ))
//Ran on leetcode:Yes
//Solution with comments:

class Solution {
    public List<String> addOperators(String num, int target) {
        backtrack(num, target,0, 0,0,"");//starting with index 0
        return output;
    }
    List<String> output= new ArrayList<>();
    
    public void backtrack(String num, int target, int index, long prevSum , long prevElem, String exp){
        
        if(index>=num.length()){//we only check results when we reach end of string
            if(prevSum==target){
                output.add(exp);

            }
            return;
        }

        for(int i=index;i<num.length();i++){
            Long currElem= Long.parseLong(num.substring(index,i+1));// ceate substring at each point so that we also have                                                                                  combination of multiple character
            
            if(num.charAt(index)=='0' && i!=index)
                continue;
            
            if(index==0)
                backtrack(num, target, i+1, currElem, currElem, exp+currElem);//do no operation when we are at starting index
            else{
                //+
                backtrack(num, target, i+1, prevSum+currElem, currElem, exp+"+"+currElem );

                //-
                backtrack(num, target, i+1, prevSum-currElem, currElem*-1, exp+"-"+currElem);

                //*
                backtrack(num, target, i+1 , (prevSum-prevElem)+(currElem*prevElem), prevElem*currElem, exp +"*"+currElem );
            }
            
        }
        
    }
}

File: Probelm2.py
class Solution(object):
    def addOperators(self, num, target):
        """
        Time complexity - 4^n
        Space complexity - 4^n , n --> length of the input number
        :type num: str
        :type target: int
        :rtype: List[str]
        """
        def backtrack(index, expression, last_operand, current_value):
            # Base case: If we have reached the end of the num string,
            # check if the current_value is equal to the target value.
            if index == len(num):
                if current_value == target:
                    result.append(expression)
                return

            # Loop through all possible lengths of the operand starting from the current index.
            for i in range(index, len(num)):
                # Convert the current operand to an integer value.
                operand = int(num[index:i+1])

                # If the index is 0, no need to add any operator.
                if index == 0:
                    backtrack(i + 1, str(operand), operand, operand)
                else:
                    # Try addition: update the current_value and expression.
                    backtrack(i + 1, expression + '+' + str(operand),
                              operand, current_value + operand)

                    # Try subtraction: update the current_value and expression.
                    backtrack(i + 1, expression + '-' +
                              str(operand), -operand, current_value - operand)

                    # Try multiplication: update the current_value and expression.
                    backtrack(i + 1, expression + '*' + str(operand), last_operand *
                              operand, current_value - last_operand + last_operand * operand)

                # Handle leading zeros in operands. If the operand is 0, no more operators can be added.
                if operand == 0:
                    break

        result = []  # Initialize an empty list to store all valid expressions.
        # Start the recursive backtracking from index=0 with empty expression.
        backtrack(0, '', 0, 0)
        return result
